crates/adapters/src/integrated/postgres/output.rs

```diff
@@ -0,0 +1,299 @@
+            .ok_or(ControllerError::not_supported(
+                "postgres output connector requires index or primary key definition",
+            ))?;
+
+        let insert = {
+            let insert = format!(
+                "INSERT INTO {table} SELECT * FROM jsonb_populate_recordset(NULL::{table}, $1::jsonb)"
+            );
+            client
+                .prepare_typed(&insert, &[postgres::types::Type::VARCHAR])
+                .unwrap_or_else(|e| panic!("failed to prepare insert statement: `{insert}`: {e}"))
+        };
+
+        let delete = {
+            let (table_keys, d_keys): (Vec<_>, Vec<_>) = keys
+                .iter()
+                .map(|k| (format!("{table}.{k}"), format!("d.{k}")))
+                .unzip();
+
+            let delete = format!(
+                "DELETE FROM {table} USING (SELECT {} FROM jsonb_populate_recordset(NULL::{table}, $1::jsonb)) as d where ({}) = ({})",
+                keys.iter().map(|k| k.as_str()).collect::<Vec<_>>().join(", "),
+                table_keys.join(", "),
+                d_keys.join(", "),
+
+            );
+            client
+                .prepare_typed(&delete, &[postgres::types::Type::VARCHAR])
+                .unwrap_or_else(|e| panic!("failed to prepare delete statement: `{delete}`: {e}"))
+        };
+
+        let out = Self {
+            endpoint_id,
+            endpoint_name: endpoint_name.to_owned(),
+            table,
+            start: Instant::now(),
+            client,
+            transaction: None,
+            insert,
+            delete,
+            keys,
+            _pin: PhantomPinned,
+        };
+
+        let _guard = out.span();
+
+        Ok(out)
+    }
+
+    fn transaction(&mut self) -> AnyResult<&mut postgres::Transaction<'static>> {
+        self.transaction.as_mut().ok_or(anyhow!("postgres: unreachable: attempting to perform a transaction that hasn't been created yet"))
+    }
+
+    pub fn insert(&mut self, value: &str) -> AnyResult<()> {
+        if value.len() <= 2 {
+            return Ok(());
+        }
+
+        let ins = self.insert.clone();
+        self.transaction()?.execute(&ins, &[&value])?;
+        Ok(())
+    }
+
+    pub fn delete(&mut self, value: &str) -> AnyResult<()> {
+        if value.len() <= 2 {
+            return Ok(());
+        }
+
+        let del = self.delete.clone();
+        self.transaction()?.execute(&del, &[&value])?;
+        Ok(())
+    }
+
+    pub fn span(&self) -> EnteredSpan {
+        info_span!(
+            "postgres_output",
+            ft = false,
+            id = self.endpoint_id,
+            name = self.endpoint_name,
+            pg_table = self.table,
+        )
+        .entered()
+    }
+}
+
+impl OutputConsumer for PostgresOutputEndpoint {
+    fn max_buffer_size_bytes(&self) -> usize {
+        usize::MAX
+    }
+
+    fn batch_start(&mut self, _: Step) {
+        let transaction: postgres::Transaction<'static> = unsafe {
```

Comment:
Add a comment explaining why this is safe.

Reply:
Do we need to add a Drop implementation to `PostgresOutputEndpoint` to make sure that the transaction is dropped before the client?

Reply:
Shouldn't `batch_end` always be called before `PostgresOutputEndpoint` is dropped?

But yeah, implementing Drop should be better.